src/node_crypto.cc

```diff
@@ -4556,6 +4575,43 @@ bool ECDH::IsKeyPairValid() {
 }
 
 
+struct CryptoJob : public ThreadPoolWork {
+  Environment* const env;
+  std::unique_ptr<AsyncWrap> async_wrap;
+  inline explicit CryptoJob(Environment* env) : ThreadPoolWork(env), env(env) {}
+  inline void AfterThreadPoolWork(int status) final;
+  virtual void AfterThreadPoolWork() = 0;
+  static inline void Run(std::unique_ptr<CryptoJob> job, Local<Value> wrap);
+};
```

Comment:
Is there any specific advantage to having the separate `AsyncWrap` JS class as opposed to what we currently do, making the individual request objects (in this case `CryptoJob`) inherit from `AsyncWrap`?

Reply:
Separation of concerns, mainly.  I initially mashed them together like you suggest but I didn't like how it looked and worked.

(There's a lot I don't like about how AsyncWrap currently works but I'll save that for another issue.)